### PR TITLE
chore(kuma-cp) migrate DiscoveryRequest/Response in KDS to V3

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### Summary

KDS is a protocol to exchange Kuma entities between Global and Zone CP.
It is based on xDS (we use go-control-plane), but instead of Envoy resources we put Kuma resources.
Right now we use DiscoveryRequest/Response V2, but V2 protos were removed from Go Control Plane so we need to migrate in order to upgrade Go Control Plane in the future.

The logical xDS definition looks like this
```
service KumaDiscoveryService {
  rpc StreamKumaResources(stream envoy.service.discovery.v3.DiscoveryRequest)
      returns (stream envoy.service.discovery.v3.DiscoveryResponse);
}
```

But we in case of KDS, both Global and Zone CPs are logical servers and clients. That's why we created mux.proto on top of this
```
service MultiplexService {
  rpc StreamMessage(stream Message) returns (stream Message);
}

message Message {
  oneof value {
    envoy.api.v2.DiscoveryRequest request = 1;
    envoy.api.v2.DiscoveryResponse response = 2;
  }
}
```

so on the wire, the real gRPC service is MultiplexService. Global is a real server and Zone is a real client, but logically under the hood - both are servers and clients.

Because `KumaDiscoveryService` is a logical service, I just migrated this to V3 protos.
In the real service I added new fields
```
message Message {
  oneof value {
    envoy.api.v2.DiscoveryRequest request = 1;
    envoy.api.v2.DiscoveryResponse response = 2;
    envoy.service.discovery.v3.DiscoveryRequest requestV3 = 3; # <- new
    envoy.service.discovery.v3.DiscoveryResponse responseV3 = 4; # <- new
  }
}
```

so new CP supports both V2 and V3 DiscoveryRequest.

We decide whether to send V2 or V3 proto on the wire, by a new gRPC MD header that we are sending.
Under the hood we convert between V2 and V3.

This way we can provide smooth migration. In the next major release, we can just remove the old `request`/`response`.

### Full changelog

* Implement support V3 DiscoveryRequest and DiscoveryResponse while keeping the compatibility with V2

### Issues resolved

No Github issue.

### Documentation

- [X] No docs

### Testing

- [ ] Unit tests
- [X] E2E tests (added a test with upgrade of CP)
- [X] Manual testing on Universal
- [ ] Manual testing on Kubernetes 

### Backwards compatibility

- [X] Add `backport-to-stable` label if the code is backwards compatible. Otherwise, list breaking changes.

### CODE REVIEW

File: api/mesh/v1alpha1/mux.proto [hunk 1]
```diff
@@ -14,5 +15,7 @@ message Message {
   oneof value {
     envoy.api.v2.DiscoveryRequest request = 1;
     envoy.api.v2.DiscoveryResponse response = 2;
+    envoy.service.discovery.v3.DiscoveryRequest requestV3 = 3;
+    envoy.service.discovery.v3.DiscoveryResponse responseV3 = 4;
```
Comment (lahabana) [inline api/mesh/v1alpha1/mux.proto:19]: I don't think the proto fields are in the message at all would it make sense to rename the field properly? Have the v3 be `request/response` and the v2 by `deprecated_request` or `legacy_` ?

Or simply put v2 too?

File: pkg/kds/client/sink_test.go [hunk 2]
```diff
@@ -37,18 +37,18 @@ var _ = Describe("KDS Sink", func() {
 
 	It("", func() {
```
Comment (lahabana) [inline pkg/kds/client/sink_test.go:38]: nit: Can we add a string to this `It` while we're at it :)

File: test/e2e/compatibility/cp_compatibility_kubernetes_multizone.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	. "github.com/onsi/gomega"
+
+	"github.com/kumahq/kuma/pkg/config/core"
+	. "github.com/kumahq/kuma/test/framework"
+)
+
+var OldChart = "0.6.3"
+var UpstreamImageRegistry = "kumahq"
+
+func CpCompatibilityMutltizoneKubernetes() {
```
Comment (lahabana) [inline test/e2e/compatibility/cp_compatibility_kubernetes_multizone.go:20]: nit: type

**Review by lahabana — APPROVED** [review]

File: api/mesh/v1alpha1/mux.proto (hunk 1 above)
Comment (jakubdyszkiewicz) [reply api/mesh/v1alpha1/mux.proto]: Good point! I forgot that we don't convert this to JSON/YAML. Changed the old field to legacyRequest

**Pull request merged**